### PR TITLE
Handle non finite edge cases with pluralization()

### PR DESCRIPTION
**Pull request by rundel**
This is a draft resolution to #701 and some related edge cases. The issue seems to be more general than originally described with handling any of the non-finite values (including NAs).

The currently implemented behavior is the following:

* Numeric non-finite values (`NA_real_`, `NA_integer_`, `NaN`, `Inf` and `-Inf`) will be treated as a `qty` > 1 for pluralization purposes.

* Non-numeric NA values (`NA_character_` and `NA`) are treated as a non-numeric input so `qty` is based on the vector's length.

I've opted not to throw an error for numeric `NA`s or `NaN` cases since that seemed a bit severe but it also seems like a plausible outcome and happy to switch to that behavior if that is a better option.

### CODE REVIEW

File: tests/testthat/test-pluralization.R [hunk 1]
```diff
@@ -103,3 +103,44 @@ test_that("issue 158", {
     print(pluralize("{9} word{?A/B/}"))
   })
 })
+
+test_that("issue 701", {
```
Comment (krlmlr) [inline tests/testthat/test-pluralization.R:107]: Perhaps:

```suggestion
test_that("Edge cases for pluralize() (#701)", {
```

?

File: tests/testthat/_snaps/pluralization.md [hunk 2]
```diff
@@ -176,3 +176,96 @@
     Output
       9 word
 
+# issue 701
+
+    Code
+      print(pluralize("{NA} file{?s} expected"))
+    Output
+      NA file expected
```
Comment (krlmlr) [inline tests/testthat/_snaps/pluralization.md:184]: Should we literally do `NA`, or something else?

```suggestion
      ? file expected
```

File: tests/testthat/_snaps/pluralization.md [hunk 3]
```diff
@@ -176,3 +176,96 @@
...
+    Output
+      Will remove the NA package.
+    Code
+      print(pluralize("Will remove {?no/the/the} {NA_character_} package{?s}."))
+    Output
+      Will remove the NA package.
+    Code
+      print(pluralize("Will remove {?no/the/the} {NA_real_} package{?s}."))
+    Output
+      Will remove the NA packages.
```
Comment (krlmlr) [inline tests/testthat/_snaps/pluralization.md:254]: Why is this different from the other two outputs above?

```suggestion
      Will remove the NA package.
```

**Review by krlmlr — COMMENTED** [review]
Thanks! The commit suggestions are just ideas and by no means authoritative.

File: tests/testthat/_snaps/pluralization.md (hunk 3 above)
Comment (rundel) [reply tests/testthat/_snaps/pluralization.md:254]: There is earlier code that distinguishes between numeric and non-numeric values. cli assumes that numeric values are a quantity while everything else is a vector where the length is the quantity. Seemed safer to keep that logic at that level even if the behavior is a bit weird.

File: tests/testthat/_snaps/pluralization.md (hunk 2 above)
Comment (hadley) [reply tests/testthat/_snaps/pluralization.md:184]: This might need to wait for @gaborcsardi, because I wonder if missing values should be an error?

File: tests/testthat/_snaps/pluralization.md (hunk 3 above)
Comment (rundel) [reply tests/testthat/_snaps/pluralization.md:254]: This also is a bit weird based on the behavior noted in #717, because a vector with an NA anywhere gets collapsed to just `NA` by `ansi_collapse()` which makes things even more confusing.

File: tests/testthat/_snaps/pluralization.md (hunk 2 above)
Comment (gaborcsardi) [reply tests/testthat/_snaps/pluralization.md:184]: I think we can process missing values, I would consider that a feature. Calling them NA is fine as well, people using R should be used to that.

**Comment by gaborcsardi** [comment]
Looks great, it is exactly the solution I had in mind. :D
